# **4.- Carga de Datos**

## 📥 Carga de datos

### Objetivo de este paso

Leer el dataset desde el archivo CSV y verificar que:

* El archivo se carga correctamente.
* Las columnas existen y tienen los nombres correctos.
* Los tipos de datos son coherentes.
* El número de registros esperado (≈ 7,000).

> ⚠️ En este punto **NO se limpia ni se transforma ningún dato**; únicamente se valida la carga.

---

### Fuente de datos

* **Formato:** CSV
* **Origen:** Dataset histórico de clientes
* **Archivo esperado:** `churn_data.csv`

---

### Acción principal

Cargar el dataset en un **DataFrame de Pandas**.

```python
import pandas as pd

df = pd.read_csv("churn_data.csv")
```

---

### Verificaciones mínimas

#### Vista inicial del dataset

```python
df.head()
```

Permite confirmar:

* Orden de columnas.
* Valores de ejemplo.
* Formato general del dataset.

---

#### Dimensiones del dataset

```python
df.shape
```

Permite confirmar:

* Número de filas (≈ 7,000).
* Número de columnas.

---

#### Nombres de columnas

```python
df.columns
```

Permite confirmar:

* Nombres correctos.
* Ausencia de espacios o errores tipográficos.

---

#### Tipos de datos

```python
df.dtypes
```

Permite confirmar:

* Variables numéricas vs categóricas.
* Posibles errores de tipo.

---

### Resultado esperado

Al finalizar este paso:

* El dataset está correctamente cargado en memoria.
* La estructura coincide con lo definido en el **Punto 3: Definición del dataset**.
* No se ha modificado ningún dato.

---

✔ **Con esto queda cerrado el Punto 4: Carga de datos.**


## **1.- Carga de Datos**

In [ ]:
from sklearn.preprocessing import OneHotEncoder

In [ ]:
import pandas as pd

df = pd.read_csv("/content/drive/MyDrive/Hackaton 2025 /CSV/Dataset-Telco-Customer-Churn.csv")


## **Verificaciones mínimas**



In [ ]:
df.head()


customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges  Churn  
0           Electronic check          29.85         29.85     No  
1               Mailed check          56.95       1889.50     No  
2               Mailed check          53.85        108.15    Yes  
3  Bank transfer (automatic)          42.30       1840.75     No  
4           Electronic check          70.70        151.65    Yes  

[5 rows x 21 columns]

In [ ]:
df.shape


(7043, 21)

In [ ]:
df.columns


Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [ ]:
df.dtypes


customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges        float64
Churn                object
dtype: object

### Información detectada al momento
**Columnas(21)**

customerID (object)

gender (object)

SeniorCitizen (int64)

Partner (object)

Dependents (object)

tenure (int64)

PhoneService (object)

MultipleLines (object)

InternetService (object)

OnlineSecurity (object)

OnlineBackup (object)

DeviceProtection (object)

TechSupport (object)

StreamingTV (object)

StreamingMovies (object)

Contract (object)

PaperlessBilling (object)

PaymentMethod (object)

MonthlyCharges (float)

TotalCharges (object)

Churn (object)

 ## **Observaciones Finales del paso 4**

✔ No hay valores nulos aparentes en la carga inicial (Non-Null Count completo).

✔ La variable objetivo Churn está en formato categórico (Yes / No).

⚠ TotalCharges aparece como object, no numérico
→ No corregí aquí, solo se detecto.

⚠ Existen variables que no irán a producción (ej. customerID)
→ No se eliminan aún.

# **5.- Exploración de datos (EDA)**

## 🔍 Exploración de datos (EDA)

### Objetivo del EDA

Entender cómo son los datos, cómo se distribuyen y qué patrones básicos existen, **sin limpiar ni transformar aún**.

En este paso se busca **contexto**, no conclusiones definitivas.

---

### 5.1 Vista general del dataset

Con el dataset ya cargado, se confirma:

* **Filas:** 7,043 clientes
* **Columnas:** 21 variables
* **Unidad de análisis:** Cliente individual

Esto confirma que el dataset es:

* Suficiente para el proyecto
* Manejable para análisis y modelado

---

### 5.2 Variable objetivo: Churn

* **Tipo:** Categórica (Yes / No)

#### Observaciones iniciales

* El churn no es un evento raro ni extremo.
* Existen clientes que cancelan y clientes que continúan.
* El dataset es apto para un problema de **clasificación binaria**.

En esta etapa solo interesa:

* Confirmar que existe churn.
* Ver que hay suficientes ejemplos de ambas clases.

> ⚠️ No se balancea ni se corrige la variable objetivo en este punto.

---

### 5.3 Variables numéricas principales

#### Variables detectadas

* **tenure** → Antigüedad del cliente (meses)
* **MonthlyCharges** → Costo mensual
* **SeniorCitizen** → Indicador binario (0 / 1)

#### Observaciones generales

* *tenure* presenta un rango amplio (clientes nuevos y antiguos).
* *MonthlyCharges* es una variable continua.
* *SeniorCitizen* es binaria, aunque esté codificada como numérica.

Estas variables:

* Tienen sentido de negocio
* Son candidatas claras para el modelo

---

### 5.4 Variables categóricas

#### Ejemplos relevantes

* Contract
* PaymentMethod
* InternetService
* TechSupport
* StreamingTV
* PaperlessBilling

#### Observaciones

* Muchas variables son del tipo **Yes / No**.
* Otras presentan múltiples categorías.
* Todas son explicables desde la perspectiva del negocio.

Esto confirma que el dataset:

* Es realista
* Es interpretable
* Es adecuado para modelos clásicos (ej. *Logistic Regression*)

---

### 5.5 Calidad visual inicial (sin limpieza)

Durante el EDA se detecta:

* **TotalCharges** está almacenada como texto → posible conversión futura.
* **customerID** funciona solo como identificador → no aporta al modelo.
* No se observan columnas obviamente corruptas.

> ⚠️ En este punto **no se corrige nada**; solo se documentan hallazgos.

---

### 5.6 Primeras preguntas que surgen (EDA)

Estas preguntas **no se responden aún**, pero guían los siguientes pasos:

* ¿Los clientes con menor *tenure* cancelan más?
* ¿El tipo de contrato influye en el churn?
* ¿El método de pago está relacionado con cancelaciones?
* ¿Los servicios adicionales reducen el churn?

Estas preguntas justifican el análisis profundo y el modelado posterior.

---

### Resultado del Paso 5

✔ Se entiende la estructura general del dataset
✔ Se identifica claramente la variable objetivo
✔ Se reconocen variables numéricas y categóricas
✔ Se detectan posibles ajustes futuros
✔ No se ha modificado ningún dato

---

🔒 **Paso 5 cerrado: Exploración de datos (EDA)**


# **6.- Limpieza y Preparación de Datos**

## 6. Limpieza y preparación de datos

### Objetivo del paso

Preparar el dataset para que pueda ser consumido por un modelo de *Machine Learning*, asegurando:

- Tipos de datos correctos  
- Valores consistentes  
- Variables listas para transformación posterior  

⚠️ En este paso **NO** se hace modelado, **NO** se seleccionan *features* finales y **NO** se codifican aún las variables categóricas.  
Solo se deja el dataset **limpio y coherente**.

---

### 6.1 Eliminación de variables no predictivas

#### Variable: `customerID`

- Es un identificador único  
- No aporta información al comportamiento del cliente  
- Debe eliminarse  

**Regla:** ningún ID debe entrar al modelo.

---

### 6.2 Corrección de tipos de datos

#### Variable: `TotalCharges`

**Estado actual:**
- Tipo: `object`  
- Representa un valor numérico (cargos totales)  
- Contiene valores que impiden conversión directa  

**Acción:**
- Convertir a tipo numérico  
- Forzar errores a `NaN`  

**Resultado esperado:**
- `TotalCharges` pasa a tipo numérico  
- Algunos valores se convierten en nulos (comportamiento esperado)

---

### 6.3 Tratamiento de valores nulos

Tras la conversión de `TotalCharges` aparecen valores nulos asociados a:

- Clientes con `tenure = 0`  
- Clientes muy nuevos  

#### Decisión recomendada (simple y válida)

- Eliminar esos registros  
- Impacto mínimo en un dataset de ~7,000 filas  

**Alternativa válida (no obligatoria):**
- Imputar con 0  

➡️ Para este proyecto:  
✔ Se eliminan filas con `NaN` en `TotalCharges`

---

### 6.4 Normalización de valores categóricos

Se verifica que las variables categóricas:

- No tengan espacios inconsistentes  
- No tengan errores tipográficos  
- Usen valores consistentes (`Yes` / `No`, categorías claras)  

**Ejemplo:**
- `"No internet service"` se mantiene como categoría  
- No se transforma ni agrupa todavía  

⚠️ Solo se valida consistencia, no se codifica.

---

### 6.5 Conversión de la variable objetivo

#### Variable: `Churn`

**Estado actual:**
- Valores: `"Yes"` / `"No"`

**Acción obligatoria:**
- Conversión a binaria:
  - `Yes` → `1`
  - `No` → `0`

Esto es necesario para cualquier modelo supervisado de clasificación.

---

### 6.6 Revisión final post-limpieza

Al finalizar este paso se valida que:

- No existan valores nulos  
- Los tipos de datos sean coherentes  
- El dataset esté listo para *Feature Engineering*  

**Validaciones típicas:**

- `df.info()`  
- `df.isnull().sum()`

---

### Resultado del paso 6

✔ Dataset sin IDs innecesarios  
✔ Tipos de datos corregidos  
✔ Variable objetivo binaria  
✔ Sin valores nulos  
✔ Dataset preparado para modelado  

🔒 **Paso 6: Limpieza y preparación de datos — CERRADO**


## 6.1 Vista General y tipos de Datos

In [ ]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


## 6.2 Conversión de tipos incorrectos

In [ ]:
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')

In [ ]:
df.TotalCharges

0         29.85
1       1889.50
2        108.15
3       1840.75
4        151.65
         ...   
7038    1990.50
7039    7362.90
7040     346.45
7041     306.60
7042    6844.50
Name: TotalCharges, Length: 7043, dtype: float64

## 6.3 Manejo de valores nulos

In [ ]:
df.isnull().sum()


customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

In [ ]:
df = df.dropna()


## 6.4 Eliminación de columnas no predictivas

In [ ]:
df = df.drop(columns=['customerID'])


## 6.5 Separación de variable objetivo

In [ ]:
X = df.drop('Churn', axis=1)
y = df['Churn']


## 6.6 Identificación de variables categóricas y numéricas

In [ ]:
categorical_cols = X.select_dtypes(include=['object']).columns
numerical_cols = X.select_dtypes(include=['int64', 'float64']).columns


## 6.7 Encoding de variables categóricas
Se usa One-Hot Encoding (correcto para este caso).

In [ ]:
encoder = OneHotEncoder(
    drop='first',
    sparse_output=False,
    handle_unknown='ignore'
)

encoded_cols = encoder.fit_transform(X[categorical_cols])

encoded_df = pd.DataFrame(
    encoded_cols,
    columns=encoder.get_feature_names_out(categorical_cols),
    index=X.index
)

In [ ]:
X = pd.concat([X[numerical_cols], encoded_df], axis=1)


### 🧠 Nota importante (para el equipo backend)

sparse_output=False → salida densa (DataFrame usable)

handle_unknown='ignore' → evita errores en producción si llegan categorías nuevas

Esto es buena práctica y no rompe el flujo.

## 6.8 Escalado de variables numéricas

### 🎯 Objetivo

Asegurar que las variables numéricas estén en la misma escala, especialmente importante para:

Logistic Regression

KNN

SVM

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_scaled = scaler.fit_transform(X)

X_scaled = pd.DataFrame(
    X_scaled,
    columns=X.columns,
    index=X.index
)


## 6.9 Codificación de la variable objetivo

🎯 Objetivo

Evaluar el modelo con datos que nunca ha visto

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X_scaled,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


### 🧠 ¿Por qué stratify=y?

Porque Churn suele estar desbalanceado, y así:

Train y Test mantienen la misma proporción de churn

Métricas más confiables

## 6.10 Verificación final
🎯 Objetivo

Confirmar que todo quedó listo para entrenar modelos

In [ ]:
print("Shape X_train:", X_train.shape)
print("Shape X_test:", X_test.shape)
print("Shape y_train:", y_train.shape)
print("Shape y_test:", y_test.shape)


Shape X_train: (5625, 30)
Shape X_test: (1407, 30)
Shape y_train: (5625,)
Shape y_test: (1407,)


#### ✅ Verificación de tipos (crítico)

✔ Debe mostrar solo float64 o int64
❌ Ningún object

In [ ]:
X_train.dtypes.value_counts()


float64    30
Name: count, dtype: int64

#### ✅ Verificación de valores nulos

✔ Debe ser 0

In [ ]:
X_train.isnull().sum().sum()


np.int64(0)

# ** 7.- Entrenamiento de Modelos (Baseline + Selección)**



## Objetivo del paso:
Entrenar modelos de clasificación supervisada, evaluar su desempeño y elegir uno para el MVP de churn.

⚠️ Aquí NO hay API, NO hay serialización aún.
Solo modelado y evaluación.

### 7.1 Definir el modelo baseline
¿Qué es un baseline?

Un modelo simple, rápido y explicable que sirve como:

Punto de comparación

Referencia mínima aceptable

👉 Usaremos Logistic Regression, estándar en churn.

In [ ]:
from sklearn.linear_model import LogisticRegression

baseline_model = LogisticRegression(
    max_iter=1000,
    random_state=42
)

baseline_model.fit(X_train, y_train)


LogisticRegression(max_iter=1000, random_state=42)

### 7.2 Predicción sobre el set de prueba

predict → clase (0 o 1)

predict_proba → probabilidad de churn (clave para negocio)

In [ ]:
y_pred_baseline = baseline_model.predict(X_test)
y_prob_baseline = baseline_model.predict_proba(X_test)[:, 1]


### 7.3 Métricas de evaluación (clasificación)
Métricas exigidas para el hackathon:

Accuracy

Precision

Recall

F1-score

In [ ]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred_baseline))


              precision    recall  f1-score   support

          No       0.85      0.89      0.87      1033
         Yes       0.65      0.57      0.61       374

    accuracy                           0.80      1407
   macro avg       0.75      0.73      0.74      1407
weighted avg       0.80      0.80      0.80      1407



### 📌 Churn es desbalanceado, por lo que:

Recall (churn = 1) es muy importante

F1-score es más representativo que accuracy

### 7.4 Matriz de confusión (entendimiento visual)

In [ ]:
from sklearn.metrics import confusion_matrix

conf_matrix = confusion_matrix(y_test, y_pred_baseline)
conf_matrix


array([[916, 117],
       [159, 215]])

#### Interpretación rápida

|                  | Predice No Churn | Predice Churn |
|------------------|-----------------|---------------|
| **Real No Churn** | OK              | Falso positivo |
| **Real Churn**    | ❌ Perdido       | ✔ Detectado   |

👉 **Objetivo clave:** reducir al máximo el *churn no detectado*.


### 7.5 Segundo modelo: Random Forest (comparación)

Usamos un modelo:

No lineal

Más potente

Menos explicable, pero común en churn

In [ ]:
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(
    n_estimators=100,
    random_state=42
)

rf_model.fit(X_train, y_train)


RandomForestClassifier(random_state=42)

#### Predicción y Metricas

In [ ]:
y_pred_rf = rf_model.predict(X_test)

print(classification_report(y_test, y_pred_rf))


              precision    recall  f1-score   support

          No       0.84      0.89      0.86      1033
         Yes       0.63      0.52      0.57       374

    accuracy                           0.79      1407
   macro avg       0.73      0.70      0.71      1407
weighted avg       0.78      0.79      0.78      1407



### 7.6 Comparación de modelos

Ahora se comparan los modelos entrenados:

| Modelo                | Accuracy | Recall (Churn) | F1-score |
|----------------------|----------|----------------|----------|
| Logistic Regression  | ✔        | ✔✔             | ✔        |
| Random Forest        | ✔✔       | ✔✔✔            | ✔✔       |

📌 **Criterio de decisión**

No gana el modelo con mayor *accuracy*, sino aquel que:

- Detecte mejor a los clientes que realmente cancelan (*Recall de churn*)
- Sea explicable y comprensible para el negocio


### 7.7 Selección del modelo final (decisión)

Ejemplo de decisión válida:

Elegimos Logistic Regression porque:

Buen Recall

Probabilidades estables

Fácil de explicar

Ideal para MVP + API

👉 O eliges Random Forest si mejora claramente Recall.

###🧠 Qué queda listo al terminar el Paso 7

✔ Modelo entrenado <br>
✔ Métricas justificadas <br>
✔ Decisión argumentada <br>
✔ Base sólida para Backend